Enhancement: `restic rewrite`: allow include filters in command

wplapper opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
----------------------------------------------------- 
This enhancement enables the use of include filters for the rewrite command. Instead of using complex logic to exclude everything else bar the files and directories which the user is interested in, it can now be defined as include "all the bits I am interested in".

File internal/filter/include.go gained an "Empty() function, and file internal/walker/rewriter.go gained logic to recognize empty sub-trees and suppresses the creation for include filtering.

cmd_rewite adds code for inclusion filters. Only sub-trees which are needed will be created.
<!--
Describe the changes and their purpose here, as detailed as needed.
-->

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
closes https://github.com/restic/restic/issues/4278

<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->

Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].

Please always follow these steps:
- Read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- Enable [maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- Run `gofmt` on the code in all commits.
- Format all commit messages in the same style as [the other commits in the repository](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
-->

- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [X] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I'm done! This pull request is ready for review.

wplapper commented:
I have managed to crack the issue of empty subtrees. This change is incorporated in internal/walker/rewriter.go

MichaelEischer commented on .github/workflows/tests.yml:
```diff
@@ -259,7 +259,7 @@ jobs:
         with:
           # Required: the version of golangci-lint is required and must be specified without patch version: we always use the latest patch version.
-          version: v1.61.0
+          version: v1.63.4
```
Please rebase this PR instead of adding merge commits

MichaelEischer commented on cmd/restic/cmd_repair_snapshots.go:
```diff
@@ -122,19 +122,19 @@ func runRepairSnapshots(ctx context.Context, gopts GlobalOptions, opts RepairOpt
 			return node
 		},
-		RewriteFailedTree: func(_ restic.ID, path string, _ error) (restic.ID, error) {
+		RewriteFailedTree: func(_ restic.ID, path string, _ error) (restic.ID, *restic.SnapshotSummary, error) {
```
How did that change from https://github.com/restic/restic/pull/5185 end up here?

MichaelEischer commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -118,50 +133,97 @@ func rewriteSnapshot(ctx context.Context, repo *repository.Repository, sn *resti
+						ifun, childMayMatch := include(nodepath)
+						return ifun && childMayMatch
+					} else if node.Type == restic.NodeTypeSymlink {
+						includeLink, _ := followSymlink(nodepath, node, include)
```
Symlinks should be treated like normal files in regard to include patterns. Everything else is unusable.

MichaelEischer commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -118,50 +133,97 @@ func rewriteSnapshot(ctx context.Context, repo *repository.Repository, sn *resti
+						return includeLink
+					} else {
+						//exclude all device files, fifo, socket, ...
+						return false
```
Nope. Those must be handled like normal files. Please follow the behavior of the backup command as closely as possible. Everything else will confuse users.

MichaelEischer commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -118,50 +133,97 @@ func rewriteSnapshot(ctx context.Context, repo *repository.Repository, sn *resti
-		rewriter, querySize := walker.NewSnapshotSizeRewriter(rewriteNode)
+		//common
+		keepDirectories := len(includeByNameFuncs) == 0
+		rewriter, querySize = walker.NewSnapshotSizeRewriter(rewriteNode, keepDirectories)
```
If the include patterns match an empty directory, then that directory must be kept. This cannot be implemented as a boolean flag passed to the rewriter, but must be checked for each individual node.

To quote myself from https://github.com/restic/restic/pull/5188#pullrequestreview-2515978653 :

> The clean solution would be to extend the `TreeRewriter` options with a new `KeepEmtpyDirectory func(path string) bool` field. If it is unset, then the rewriter behaves like before. If the field is set and the filtered tree is empty then `KeepEmtpyDirectory` decides whether to keep the tree or not. If the tree should not be kept, then `RewriteTree` returns an empty restic.ID which also means that the corresponding node should be dropped.

Effectively the only two options here are adding the flag as return value to `rewriteNode` or passing a `KeepEmtpyDirectory` function to the rewriter.

MichaelEischer commented on internal/walker/rewriter.go:
```diff
@@ -86,11 +88,13 @@ type BlobLoadSaver interface {
 }
 
-func (t *TreeRewriter) RewriteTree(ctx context.Context, repo BlobLoadSaver, nodepath string, nodeID restic.ID) (newNodeID restic.ID, err error) {
+const EmptyNodeID = `{"nodes":[]}` + "\n"
```
Drop that constant you won't need it for a clean implementation.

MichaelEischer reviewed (changes requested):
A few preliminary comments before you try to further work on this PR

wplapper commented:
I put this PR on hold until PR https://github.com/restic/restic/pull/5185 is fully resolved.

MichaelEischer commented on cmd/restic/cmd_rewrite_integration_test.go:
```diff
@@ -170,5 +171,133 @@ func TestRewriteSnaphotSummary(t *testing.T) {
+	testListSnapshots(t, env.gopts, 1)
+}
+
+func TestRewriteExcludeNothing(t *testing.T) {
```
I'm not sure what this test case actually tests. Even `--exclude *` would not remove a snapshot as `KeepEmptyDirectories` is always true for excludes. Or is that what you wanted to check? Then the exclude pattern must be `*`.

MichaelEischer commented on cmd/restic/cmd_rewrite_integration_test.go:
```diff
@@ -170,5 +171,133 @@ func TestRewriteSnaphotSummary(t *testing.T) {
+		IncludePatternOptions: filter.IncludePatternOptions{Includes: []string{"JohannSebastianBach"}},
+		ExcludePatternOptions: filter.ExcludePatternOptions{Excludes: []string{"WolfgangAmadeusMozart"}},
+	}, env.gopts, []string{"latest"})
+	rtest.Assert(t, err != nil, "You cannot specify include and exclude options simultaneously!")
```
I'd prefer `"exclude and include patterns are mutually exclusive"` similar to the restore command as error message.

MichaelEischer commented on cmd/restic/cmd_rewrite_integration_test.go:
```diff
@@ -170,5 +171,133 @@ func TestRewriteSnaphotSummary(t *testing.T) {
+	}, env.gopts, []string{"latest"})
+	rtest.Assert(t, err != nil, "You cannot specify include and exclude options simultaneously!")
+
+	err = runRewrite(context.TODO(), RewriteOptions{
```
This should work IMHO.

MichaelEischer commented on cmd/restic/cmd_rewrite_integration_test.go:
```diff
@@ -170,5 +171,133 @@ func TestRewriteSnaphotSummary(t *testing.T) {
+			count++
+		}
+	}
+	rtest.Assert(t, count == 1, "there should be 1 empty directory in the snapshot, but there are %d snapshots", count)
```
```suggestion
	rtest.Assert(t, count == 1, "there should be 1 empty directory in the snapshot, but there are %d directories", count)
```

MichaelEischer commented on cmd/restic/cmd_rewrite_integration_test.go:
```diff
@@ -170,5 +171,133 @@ func TestRewriteSnaphotSummary(t *testing.T) {
+	rtest.Assert(t, snapshots[0] != newSnapshots[0], "snapshot id should have changed")
+
+	// read restic ls output and count lines which contain the string "empty-directory"
+	out := testRunLsWithOpts(t, env.gopts, LsOptions{}, []string{newSnapshots[0].String()})
```
Nice approach

MichaelEischer commented on cmd/restic/cmd_rewrite_integration_test.go:
```diff
@@ -170,5 +171,133 @@ func TestRewriteSnaphotSummary(t *testing.T) {
+	sn, err := restic.LoadSnapshot(context.TODO(), repo, newSnapshots[0])
+	rtest.OK(t, err)
+
+	rtest.Assert(t, sn.Summary != nil, "snapshot should have a summary attached")
```
Please additionally use `testRunLsWithOpts` to check that the directory has actually be removed.

MichaelEischer commented on cmd/restic/cmd_rewrite_integration_test.go:
```diff
@@ -170,5 +171,133 @@ func TestRewriteSnaphotSummary(t *testing.T) {
+
+	// get snapshot summary and find this 1 directory
+	_, repo, unlock, err := openWithExclusiveLock(context.TODO(), env.gopts, false)
+	rtest.OK(t, err)
```
```suggestion
	rtest.OK(t, err)
	defer unlock()
```

MichaelEischer commented on cmd/restic/cmd_rewrite_integration_test.go:
```diff
@@ -170,5 +171,133 @@ func TestRewriteSnaphotSummary(t *testing.T) {
+	rtest.Assert(t, sn.Summary != nil, "snapshot should have a summary attached")
+	rtest.Assert(t, sn.Summary.TotalFilesProcessed == 0,
+		"there should be 0 files in the snapshot, but there are %d files", sn.Summary.TotalFilesProcessed)
+	unlock()
```
```suggestion
```

MichaelEischer commented on cmd/restic/cmd_rewrite_integration_test.go:
```diff
@@ -170,5 +171,133 @@ func TestRewriteSnaphotSummary(t *testing.T) {
+	rtest.Assert(t, snapshots[0] != newSnapshots[0], "snapshot id should have changed")
+
+	// read restic ls output and count .txt files
+	out := testRunLsWithOpts(t, env.gopts, LsOptions{}, []string{newSnapshots[0].Str()})
+	fileList := strings.Split(string(out), "\n")
+	count := 0
+	for _, filename := range fileList {
+		if strings.Contains(filename, ".txt") {
+			count++
+		}
+	}
+	rtest.Assert(t, count == 2, "there should be 2 txt files in the snapshot, but there are %d files", count)
```
Please extract this into a helper function. And also use it for `TestRewriteEmptyDirectory`.

MichaelEischer commented on cmd/restic/cmd_rewrite_integration_test.go:
```diff
@@ -170,5 +171,133 @@ func TestRewriteSnaphotSummary(t *testing.T) {
+	rtest.Assert(t, count == 2, "there should be 2 txt files in the snapshot, but there are %d files", count)
+
+	// get snapshot summary and find these 2 files
+	_, repo, unlock, err := openWithExclusiveLock(context.TODO(), env.gopts, false)
```
Same here

MichaelEischer reviewed (changes requested):
Thanks for the changes so far! I have added quite a few comments, but those are mostly small changes and nitpicks. The general approach now looks like what I had in mind.

wplapper committed:
restic rewrite include - based on restic 0.18.1

cmd/restic/cmd_rewrite.go:
introduction of include filters for this command:
- add include filters, add error checking code
- add new parameter 'keepEmptyDirectoryFunc' to 'walker.NewSnapshotSizeRewriter()',
  so empty directories have to be kept to keep the directory structure intact
- add parameter 'keepEmptySnapshot' to 'filterAndReplaceSnapshot()' to keep snapshots
  intact when nothing is to be included
- introduce helper function 'gatherIncludeFilters()' and 'gatherExcludeFilters()' to
  keep code flow clean

cmd/restic/cmd_rewrite_integration_test.go:
add several new tests around the 'include' functionality

internal/filter/include.go:
this is where is include filter is defined

internal/walker/rewriter.go:
- struct RewriteOpts gains field 'KeepEmtpyDirectory', which is a 'NodeKeepEmptyDirectoryFunc()'
  which defaults to nil, so that al subdirectories are kept
- function 'NewSnapshotSizeRewriter()' gains the parameter 'keepEmptyDirecoryFilter' which
  controls the management of empty subdirectories in case of include filters active

internal/data/tree.go:
gains a function Count() for checking the number if node elements in a newly built tree

internal/walker/rewriter_test.go:
function 'NewSnapshotSizeRewriter()' gets an additional parameter nil to keeps things happy

cmd/restic/cmd_repair_snapshots.go:
function 'filterAndReplaceSnapshot()' gets an additional parameter 'keepEmptySnapshot=nil'

doc/045_working_with_repos.rst:
gets to mention include filters

changelog/unreleased/issue-4278:
the usual announcement file

git rebase master -i produced this

restic rewrite include - keep linter happy

cmd/restic/cmd_rewrite_integration_test.go:
linter likes strings.Contain() better than my strings.Index() >= 0

wplapper committed:
restic rewrite integration test

convert exclusive lock on repository to 'no-lock'

MichaelEischer commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -348,3 +361,71 @@ func runRewrite(ctx context.Context, opts RewriteOptions, gopts global.Options,
+				return true
+			}
+			matched, childMayMatch := include(nodepath)
+			if matched && childMayMatch {
```
TODO

wplapper committed:
restic rewrite --include

added function Count() to the *TreeWriter methods

MichaelEischer committed:
minor cleanups and typos

MichaelEischer commented on cmd/restic/cmd_rewrite_integration_test.go:
```diff
@@ -195,3 +218,169 @@ func TestRewriteSnaphotSummary(t *testing.T) {
+	err := testRunRewriteWithOpts(t,
+		RewriteOptions{
+			Forget:                true,
+			ExcludePatternOptions: filter.ExcludePatternOptions{Excludes: []string{"nothing-whatsoever"}},
```
This test case doesn't test anything at the moment. It would have to test using pattern `*`. However, then the test fails as it is currently perfectly valid to just exclude all files in a snapshot. For simplicity I think we should just keep it at that and drop this testcase.

MichaelEischer committed:
rewriter: test KeepEmptyDirectory option

MichaelEischer committed:
rewrite: skip snapshot parts not matchable by include patterns

MichaelEischer reviewed (approved):
LGTM. I've added a few commits that include a bunch of smaller cleanups, add tests for `KeepEmptyDirectory` and change the node filtering to only include directories that have a chance that some file in a subdirectory matches. The latter can speed up rewrite quite a lot, when for example matching against `/data` or some other absolute path.
